src/homework3/pairs_of_elements.py

```diff
@@ -0,0 +1,25 @@
+"""Пары элементов.
+
+Дан список чисел. Посчитайте, сколько в нем пар элементов, равных друг другу.
+Считается, что любые два элемента, равные друг другу образуют одну пару,
+которую необходимо посчитать.
+Входные данные - строка из чисел, разделенная пробелами.
+Выходные данные - количество пар.
+Важно: 1 1 1 - это 3 пары, 1 1 1 1 - это 6 пар
+"""
+from math import factorial
+
+str_ = '1 abc 3 4 1 abc 1 1 abc , , , foo bar foo'
+dict_ = {}  # словарь для хранения элементов строки и их количества
+pairs = 0  # количество пар, инициализировано в 0
+# создаем словарь, хранящий элементы входной строки и их количества
+for elem in str_.split():
+    dict_[elem] = dict_.get(elem, 0) + 1
+# очевидно, что если элемент 1, то он не парный и мы его не считаем
+# а когда элементов >= 2, то количество пар - это количество сочетаний
+# элементов из n по 2. Считаем по формуле: n! / (k! * (n - k)!), где k = 2
+for key in dict_:
+    if dict_[key] >= 2:
+        pairs += factorial(dict_[key]) / (2 * factorial(dict_[key] - 2))
```

Comment:
here we can find arithmetic progression and count it without factorials. And it would not be needed to write all these comments )))

Reply:
Code rewritten without use of factorial.